### ISSUE TITLE
[32B Dense] Debug 64k increasing loss

### ISSUE THREAD

**Issue by tvosch** [OpenEuroLLM/Taskboard#381]
While training the 32B dense flag model we encountered an increasing loss around 64k:
See:
<img width="824" height="235" alt="Image" src="https://github.com/user-attachments/assets/4c55d893-c4ad-4c79-953b-c6940167ebfc" />
From: https://wandb.ai/openeurollm-project/oellm_32b_dense
For live debug session: https://github.com/OpenEuroLLM/oellm-autoexp/blob/prod/oellm_32b_dense/docs/64k-debug/DEBUG.md

# The Problem:
The training loss was going down like it should. Then around **iteration 65,000**
it turned around and started going **up**, and it keeps getting worse.

- Run: `oellm_32b_dense_prod_dataopt5_gbs4096_lr3e-4`
- Last job: `1537344` →
  [wandb](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/tjeg54n7),
  stopped 2026-08-30 22:43 at iteration **75,125**
- Training loss: **1.5434** at ~65,000 → **1.587** at 75,125

This is not a wobble. It is a steady climb that is speeding up — all of it visible
in job `1537344`
([wandb](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/tjeg54n7), which
covers 66,626 → 75,125, i.e. the whole broken stretch):

| iterations | loss change per 1000 steps | where to see it |
|---|---|---|
| 50k–56k | −0.0018 (good, going down) | job `1530865` |
| 62k–66.6k | −0.0005 (nearly stopped) | jobs `1535676`, `1536694` |
| 66.6k–71k | **+0.0023** (going up) | [`1537344`](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/tjeg54n7) |
| 71k–75k | **+0.0062** (going up faster) | [`1537344`](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/tjeg54n7) |

**Why this is a big deal:** we are only **8.4% into a 15T-token run**
(75,125 of 894,000 iterations, ~1.26T tokens). If the model gets worse the longer
we train, the other 92% cannot be run this way. Rough cost so far: ~11,000
iterations spent making it worse ≈ **185B tokens ≈ 27,500 GPU-h**.

### How we found the exact spot

Fitting the healthy part of the curve (12k–34k) and extrapolating, the run stays
on track until ~66,000 and then breaks away: +0.012 at 68k, +0.024 at 72k,
+0.038 at 74k. The break is at **~66,625**.

Held-out scores confirm the model really got worse (lower = better):

| checkpoint | score | job |
|---|---|---|
| 60,000 | 1.500890 | `1548575` |
| **64,000** | **1.495481** ← best | `1573075` |
| 68,000 | 1.511213 | `1563495` |
| 72,000 | 1.519173 | `1543640` |

Normally later = better. Here it flips after 64,000. **That flip is the bug.**

This could be due to a variety of things

- [ ] FP8
- [ ] NaN weights/activations
- [ ] Switching containers
- [ ] Turning on inter-document attention masking
- [ ] Data quality
- [ ] RMSNorm off-zero gamma

**Comment by tvosch** [comment]
# On changing objective to interdocument masking and software stack mid-run (WIP)

**Is it the software stack SWAP we did mid-run?**

  We replaced the container, Transformer Engine, megatron-core and the attention
  backend **in the middle of the run**, on 2026-08-28 at **iteration 34,455**.
  Job `1524558` ([eef62uvs](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/eef62uvs))
  is the first on the new stack; `1516079`
  ([uaewx25r](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/uaewx25r))
  ran to 34,450 on the old one — **overlay those two to see the boundary.**

  | | before (`1516079`) | after (`1524558`) |
  |---|---|---|
  | container | `nemo_26.04.sif` | `MegatronTraining-JUPITER-te218-fa3` |
  | Transformer Engine | 2.14 | 2.18 |
  | megatron-core | 0.16 | 0.19 |
  | `attention_backend` | `auto` (cuDNN) | `flash` (FA3) |
  | `dataloader_inter_document_masking` | *did not exist* | **`true`** |
  | `data_parallel_sharding_strategy` | `no_shard` | `optim_grads_params` |

  Commits, all 2026-08-28: `cfe21f8` *Adopt document masking, FlashAttention 3 and
  the te218-fa3 container* · `a27f103` *Update Megatron-LM to 0.19* · `521a12d` /
  `bcb0797` *TE 2.18 container* · `390ee3c` *configs and scripts that validated the
  0.19 switch*

**Comment by kpoeppel** [comment]
Note that NaNs should lead to a hard failure. Interestingly, sometimes we get NaNs on the first training step (starting from a checkpoint). So there could be "silently erroneous" GPUs that "do a different computation" only visible through these NaN, as a second resume. This would be the absolute worst case, making just everything a probabilistic gamble.
Let's hope these NaN failures are errors leading to NaN's only and not different results throughout by chance.

**Comment by tvosch** [comment]
# On attention drift

<img width="1350" height="2475" alt="Image" src="https://github.com/user-attachments/assets/aee46207-a325-4121-b0bc-da119ef22239" />

**Comment by tvosch** [comment]
# On data ordering (closed)
Is it the data order? No. cont2 re-shuffled with seed 4321 and got worse anyway (+0.0032 held-out by 66,000). Job 1558724 → [921bmgjb](https://wandb.ai/openeurollm-project/oellm_32b_dense/runs/921bmgjb)

**Comment by tvosch** [comment]
# On FP8 (WIP)

**Comment by joaquinvanschoren** [comment]
Can we rule out inter-document attention masking by turning it off again and testing?

**Comment by JeniaJitsev** [comment]
@joaquinvanschoren this was already performed, with no change when switching off. In general, it seems for now there is no visible change in stats when switching the conf at 34k (dashed vertical line). Worrying is that linear activation amax fc2 grows from the start reaching very high values and continuing to grow, and also steady growth of weight amax from the start. Those are behaviors that appear from the start with old conf.

**Comment by JeniaJitsev** [comment]
@kpoeppel @tvosch @Vmjkom can we be more precise about NaNs that happened? It is not clear for now  under which conditions we observe those occurring. Any restart, or only when switching conf at 34k, or restarts that happened after 34k switch?

**Comment by tvosch** [comment]
# Data copied to fscratch
Data is also copied to fscratch here: `/e/fscratch/e-sta-openeurollm/training_v2/training/collection/flag`
With cache here ` /e/fscratch/e-sta-openeurollm/production_training/oellm_32b_dense/cache`
With datamix here `/e/project1/e-sta-openeurollm/production_training/oellm_32b_dense/flagship_datamix_option5_fscratch.txt`

**Comment by joaquinvanschoren** [comment]
Did we try switching FP8 off and logging the BF16/FP32 stats? I'm wondering whether it could be the FP8 scaling setup, e.g. amax history length being too short. If scaling is off the optimizer might push the weights into weird ranges.

**Comment by tvosch** [comment]
BF16 comparative run still training: https://wandb.ai/openeurollm-project/oellm_32b_dense

**Comment by tvosch** [comment]
Pushed new changes: https://github.com/OpenEuroLLM/oellm-autoexp/tree/prod/oellm_32b_dense

**Comment by Vmjkom** [comment]
> [@kpoeppel](https://github.com/kpoeppel) [@tvosch](https://github.com/tvosch) [@Vmjkom](https://github.com/Vmjkom) can we be more precise about NaNs that happened? It is not clear for now under which conditions we observe those occurring. Any restart, or only when switching conf at 34k, or restarts that happened after 34k switch?

My NaN's, as well as Korbinian's i think, happened on specific nodes and those nodes we're consequently added to the exclude list. I noticed NaN's when i was doing small time experiments starting from scratch, so not loading from a checkpoint.

**Comment by tvosch** [comment]
### On the optimizer state and effective step size (Closed as the trigger; a real mis-setting)

`scripts/scan_optimizer_state.py`
reads the fp32 Adam buffers straight out of the checkpoints (CPU only, ~2 min each).

**The quantity that matters** is not `sqrt(v)` itself but how much of each step
`adam_eps` eats:

```
step = lr * m / (sqrt(v) + eps)      suppression = eps / (sqrt(v) + eps)
```

At `sqrt(v) = eps` a parameter moves at half speed. Adam is otherwise
scale-invariant (`m/sqrt(v)` is dimensionless); `eps` is the one term that breaks
that, because it is an absolute number in gradient units.

 **Two measurement bugs, both fixed; all numbers below are re-derived.**
(a) Shards are keyed `...group_idx_G.gbuf_idx_B.dtype_(...).bucket_idx_N` and the
scan keyed on `bucket_idx` alone — 64 distinct `(group, gbuf)` shards collided in
a dict and one arbitrary survivor won, so "5 buckets" was 4.4% of the model chosen
by iteration order. (b) The per-iteration figure was an *unweighted median across
shards*, letting a 135M-parameter shard speak for 33.9B. Fixed: full-prefix shard
identity, `--spread` sampling, size-weighted aggregation.

<img width="1800" height="645" alt="Image" src="https://github.com/user-attachments/assets/7edcc6e8-4581-4412-bb0f-417623fc00d5" />

**Measurement** (`--spread`, 8 shards, 3.2B params, 9.5% of the model):

| iteration | median sqrt(v) | step suppressed by | median step |
|---|---|---|---|
| 44,000 | 5.59e-08 | 15.2% | 0.125 |
| 48,000 | 5.92e-08 | 14.5% | 0.135 |
| 52,000 | 5.39e-08 | 15.7% | 0.125 |
| 56,000 | 5.23e-08 | 16.1% | 0.124 |
| 60,000 | 5.43e-08 | 15.6% | 0.124 |
| **64,000** | 5.78e-08 | **14.8%** | 0.119 |
| 68,000 | 7.82e-08 | 11.3% | 0.138 |
| 72,000 | 8.08e-08 | 11.0% | 0.134 |
| 75,126 | 9.83e-08 | 9.2% | 0.150 |

Earlier trajectory (4.4% scan, re-weighted): suppression 6.1% at 8,000 → 10.8% at
16,000 → 14.5% at 24,000 → 15.7% at 34,454, then flat.

**Verified at 3.7x the coverage** (28 shards, 11.9B params, 35% of the model):

| | 44,000 | 60,000 |
|---|---|---|
| median sqrt(v) | 5.14e-08 | 4.95e-08 |
| suppression | 16.3% | 16.8% |
| median step | 0.1181 | 0.1178 |

**Why this is not the trigger — four independent reasons.**

1. **It saturates 35,000 iterations early.** Suppression climbs 6% → ~15% over the
   first 24,000 iterations and then moves about one point across the next 40,000.
   The run trained healthily at 15% and degraded at 15%.
2. **It is flat *through* the onset.** Damage begins ~60,500 (item 27). Between
   44,000 and 60,000 the median step changes by **0.3%** at 35% coverage.
3. **The rise comes after the loss does.** `sqrt(v)` is flat to 64,000, then climbs
   70% by 75,126 — *unwinding* the damping (15% → 9%). That is the signature of a
   consequence: a degrading model produces larger gradients, which grow `v`.
4. **Fine resolution and the bf16 control agree.** `cont3` straddles the crossing
   (65,500 → 66,625): `sqrt(v)` 4.97e-08 → 5.29e-08, step 0.0852 → 0.0851.
   `cont4b` (bf16) sits at the same suppression as the flagship (17.4% vs 16.2% at
   64,000) and degrades too — a shared quantity cannot be what separates two runs
   that behave differently.

**What is real, and worth fixing in the next run.** `adam_eps = 1e-8` is Megatron's
default, not a value chosen for this gradient scale. With a 16.8M-token batch the
per-weight gradient is ~1e-7, so 1e-8 is not negligible: **the whole model has run
~15% slow since iteration ~24,000.** The consequence that actually bites is on
weight decay. Megatron uses decoupled AdamW:

```
w <- w - lr*m/(sqrt(v)+eps) - lr*wd*w
```

The decay term is **not** divided by `sqrt(v)` and **not** touched by `eps`.
Throttling the learning term by 15% while decay keeps full strength runs the
decay-to-learning ratio **~18% above** what `wd = 0.05` was meant to set. That is
the missing link to item 16. Fix: `adam_eps = 1e-15`, as large-model recipes use.

**Conclusion — and what it says about where to look next.**

The optimizer state contains **no early warning**. Every Adam quantity we can read
is either flat through the onset or turns after the loss does. Adam state is a
mirror of the loss here, not a driver.

That is worth more than one closed candidate, because it rules out a whole *class*
of hypothesis: *"something accumulated in training state and crossed a threshold
around 60,000."* The optimizer is where such an accumulation would have to show up,
and it does not — nor do the norm gains (item 12), the embeddings, or attention
(items 18, 30). Every slowly-drifting quantity we have measured saturated long
before the onset, exactly as `eps` suppression did.

What survives is the opposite shape: settings **mis-specified from step 0 and never
corrected** — constant `lr = 3e-4` (1.7x the scaling-law optimum at batch 4096), no
batch ramp, `adam_eps = 1e-8`, `wd_mult: 0.0` on the gains. These need no event at
60,000; they only need the model to reach the point where a too-aggressive constant
setting stops paying off. That is item 29, and this item is now a second,
independent argument for it.

**Method note.** Two aggregation choices in this item each turned a flat effect into
a dramatic one: "% of parameters within 10x of eps" (reported as 7% → 89%) is a
count either side of an arbitrary cutoff, and the unweighted shard median let 135M
parameters stand for 33.9B. Compute the quantity that acts on the weights — here
`eps/(sqrt(v)+eps)` — and weight it by what it acts on.

**Comment by tvosch** [comment]
# On Deferred LM Head Wgrad enabled
https://github.com/OpenEuroLLM/oellm-autoexp/blob/prod/oellm_32b_dense/docs/64k-debug/DEBUG.md#43-deferred-lm-head-wgrad-with-overlapped-gradient-reduction

Language-based downstream evals
<img width="2400" height="825" alt="Image" src="https://github.com/user-attachments/assets/14a0d998-29e2-426b-8446-759263c89204" />
